pkg/controller/workload/workload_processor.go

```diff
@@ -501,8 +580,42 @@ func (p *Processor) storeServiceData(serviceName string, waypoint *workloadapi.G
 		newValue.EndpointCount = oldValue.EndpointCount
 	}
 
-	if err = p.bpf.ServiceUpdate(&sk, &newValue); err != nil {
-		log.Errorf("Update Service failed, err:%s", err)
+	// if it is a policy update
+	if newValue.LbPolicy != oldValue.LbPolicy {
+		if newValue.LbPolicy == uint32(workloadapi.LoadBalancing_UNSPECIFIED_MODE) { // from locality loadbalance to random
+			// In locality load balancing mode, the workloads are stored according to the calculated corresponding priorities.
+			// When switching from locality load balancing mode to random, we first update the endpoint map, as at this point,
+			// there might not be any workload with the highest priority, and directly switching the service's LB policy could
+			// lead to unexpected service disruptions.
+			if err = p.updateEndpoint(sk.ServiceId, false); err != nil { // this will change bpf map totally
+				log.Errorf("UpdateEndpoint failed, err:%s", err)
+			}
+			updateValue := bpf.ServiceValue{}
+			if err = p.bpf.ServiceLookup(&sk, &updateValue); err != nil {
+				log.Warnf("Lookup Service failed, err:%s", err)
+			}
+			updateValue.LbPolicy = newValue.LbPolicy
```

Comment:
donot we need to update the EndpointCount?

Reply:
The EndpointCount is already the newest.
In the current implementation, p.updateEndpoint synchronously maintains the state of the bpfmap to the latest when updating an endpoint. We should maintain the independence of the updateEndpoint function in managing the bpfmap state (since we currently lack a mechanism to ensure consistency between user space and kernel space, which we might consider implementing later). If the serviceValue passed internally and externally to the updateEndpoint function is the same, it could lead to complex state changes during the update process, increasing the risk of state confusion errors.